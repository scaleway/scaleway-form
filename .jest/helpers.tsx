import { ThemeProvider } from '@emotion/react'
import makeHelpers from '@scaleway/jest-helpers'
import { theme as lightTheme } from '@scaleway/ui'
import { render } from '@testing-library/react'
import type { ComponentProps, ReactElement, ReactNode } from 'react'
import { Form } from '../src/components'
import { mockErrors } from '../src/mocks'

type WrapperProps = {
  theme?: typeof lightTheme
  children: ReactNode
}

const Wrapper = ({ theme = lightTheme, children }: WrapperProps) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

export const {
  renderWithTheme,
  shouldMatchEmotionSnapshot,
  shouldMatchEmotionSnapshotWithPortal,
} = makeHelpers(Wrapper, { classNameReplacer: className => className })

export const shouldMatchEmotionSnapshotFormWrapper = (
  children: ReactNode,
  options?: Parameters<typeof shouldMatchEmotionSnapshot>[1],
  formOptions?: Partial<ComponentProps<typeof Form>>,
) => {
  const { initialValues } = formOptions ?? {}

  return shouldMatchEmotionSnapshot(
    <Form
      onRawSubmit={() => {}}
      errors={mockErrors}
      initialValues={initialValues}
    >
      {() => children}
    </Form>,
    options,
  )
}

export const renderWithWrapper = (children: ReactElement) =>
  render(children, {
    wrapper: () => (
      <ThemeProvider theme={lightTheme}>
        <Form onRawSubmit={() => {}} errors={mockErrors}>
          {children}
        </Form>
      </ThemeProvider>
    ),
  })

export const mockRandom = () =>
  jest.spyOn(global.Math, 'random').mockReturnValue(0.4155913669444804)

export const restoreRandom = () =>
  jest.spyOn(global.Math, 'random').mockRestore()
