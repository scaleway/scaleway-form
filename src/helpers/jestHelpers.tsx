import { ThemeProvider } from '@emotion/react'
import makeHelpers from '@scaleway/jest-helpers'
import { theme as scwuiTheme } from '@scaleway/ui'
import { RenderResult, render } from '@testing-library/react'
import React, { FC, ReactElement, ReactNode } from 'react'
import Form from '../components/Form'
import mockErrors from '../mocks/mockErrors'

interface WrapperProps {
  theme?: typeof scwuiTheme
}

const Wrapper: FC<WrapperProps> = ({ theme = scwuiTheme, children }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

export const {
  renderWithTheme,
  shouldMatchEmotionSnapshot,
  shouldMatchEmotionSnapshotWithPortal,
} = makeHelpers(Wrapper, { classNameReplacer: className => className })

export const shouldMatchEmotionSnapshotFormWrapper = (
  children: ReactNode,
  options?: {
    transform: (node: RenderResult) => Promise<void> | void
  },
) =>
  shouldMatchEmotionSnapshot(
    <Form errors={mockErrors}>{() => children}</Form>,
    options,
  )

export const renderWithWrapper = (children: ReactElement) =>
  render(children, {
    wrapper: () => (
      <ThemeProvider theme={scwuiTheme}>
        <Form errors={mockErrors}>{children}</Form>
      </ThemeProvider>
    ),
  })

export const mockRandom = () =>
  jest.spyOn(global.Math, 'random').mockReturnValue(0.4155913669444804)

export const restoreRandom = () =>
  jest.spyOn(global.Math, 'random').mockRestore()
