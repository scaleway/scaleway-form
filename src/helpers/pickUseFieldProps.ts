import { AnyObject } from 'final-form'
import { UseFieldConfig } from 'react-final-form'
import { BaseFieldProps } from '../types'

const pickUseFieldProps = <T = AnyObject, K = AnyObject>({
  afterSubmit,
  allowNull,
  beforeSubmit,
  data,
  defaultValue,
  format,
  formatOnBlur,
  initialValue,
  isEqual,
  multiple,
  parse,
  subscription,
  validate,
  validateFields,
  value,
}: BaseFieldProps<T, K>): UseFieldConfig<T, K> => ({
  afterSubmit,
  allowNull,
  beforeSubmit,
  data,
  defaultValue,
  format,
  formatOnBlur,
  initialValue,
  isEqual,
  multiple,
  parse,
  subscription,
  validate,
  validateFields,
  value,
})

export default pickUseFieldProps
