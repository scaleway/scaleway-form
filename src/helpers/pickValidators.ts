import type { ValidatorObject, ValidatorProps } from '../types'
import validators from '../validators'
import type { ValidatorFn } from '../validators/types'

export const pickValidators = <InputValue = unknown>(args: ValidatorProps) =>
  Object.entries(args)
    .map(([key, value]) =>
      value !== undefined
        ? (validators as Record<string, ValidatorFn>)[
            key as keyof typeof validators
          ]?.(value)
        : undefined,
    )
    .filter(validator => !!validator) as ValidatorObject<InputValue>[]
