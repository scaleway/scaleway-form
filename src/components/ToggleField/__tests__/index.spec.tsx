import { ToggleField } from '..'
import { shouldMatchEmotionSnapshotFormWrapper } from '../../../../.jest/helpers'

describe('ToggleField', () => {
  test('should render correctly', () =>
    shouldMatchEmotionSnapshotFormWrapper(<ToggleField name="test" />))

  test('should render correctly disabled', () =>
    shouldMatchEmotionSnapshotFormWrapper(<ToggleField name="test" disabled />))

  test('should render correctly checked', () =>
    shouldMatchEmotionSnapshotFormWrapper(
      <ToggleField name="test" value="test" initialValue={['test']} />,
      {
        transform: node => {
          const element = node.getByRole('checkbox') as HTMLInputElement
          expect(element.checked).toBeTruthy()
        },
      },
    ))

  test('should render correctly with label and checked', () =>
    shouldMatchEmotionSnapshotFormWrapper(
      <ToggleField name="test" initialValue label="test" />,
    ))
})
