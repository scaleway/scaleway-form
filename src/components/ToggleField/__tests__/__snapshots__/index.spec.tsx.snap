// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleField should render correctly 1`] = `
<DocumentFragment>
  .cache-ou5gie {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  cursor: pointer;
}

.cache-ou5gie:active .e6hn8de3[data-disabled='false']:after {
  width: 22.04px;
}

.cache-ou5gie[aria-disabled='true'] {
  cursor: not-allowed;
  color: #c4c9d6;
}

.cache-1w4yazo {
  box-sizing: content-box;
  outline: none;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 24px;
  position: relative;
  -webkit-transition: all 300ms;
  transition: all 300ms;
  background-color: #b2b6c3;
  width: 48px;
  height: 24px;
}

.cache-1w4yazo:after {
  content: '';
  position: absolute;
  top: 4px;
  left: 5px;
  width: 16px;
  height: 16px;
  border-radius: 100px;
  background-color: #ffffff;
  -webkit-transition: all 300ms;
  transition: all 300ms;
}

.cache-1w4yazo:focus-within,
.cache-1w4yazo:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
}

.cache-1w4yazo[data-disabled='false']:active:after {
  width: 22.04px;
}

.cache-1w4yazo[data-checked='true'] {
  color: #151a2d;
  background-color: #4f0599;
}

.cache-1w4yazo[data-checked='true']:after {
  left: calc(100% - 5px);
  -webkit-transform: translateX(-100%);
  -moz-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
}

.cache-1w4yazo[data-checked='true']:focus-within,
.cache-1w4yazo[data-checked='true']:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
}

.cache-1w4yazo[data-disabled='true'] {
  background: #d4dae7;
}

.cache-1w4yazo[data-disabled='true'][data-checked='true'] {
  background: #bda5df;
}

.cache-p9zju0 {
  position: absolute;
  opacity: 0;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  cursor: pointer;
}

.cache-p9zju0[disabled] {
  cursor: not-allowed;
}

<form
    novalidate=""
  >
    <label
      aria-disabled="false"
      class="cache-ou5gie e6hn8de1"
    >
      <div
        class="cache-1w4yazo e6hn8de3"
        data-checked="false"
        data-disabled="false"
      >
        <input
          aria-checked="false"
          aria-label="test"
          class="cache-p9zju0 e6hn8de2"
          id="test"
          name="test"
          type="checkbox"
        />
      </div>
    </label>
  </form>
</DocumentFragment>
`;

exports[`ToggleField should render correctly checked 1`] = `
<DocumentFragment>
  .cache-ou5gie {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  cursor: pointer;
}

.cache-ou5gie:active .e6hn8de3[data-disabled='false']:after {
  width: 22.04px;
}

.cache-ou5gie[aria-disabled='true'] {
  cursor: not-allowed;
  color: #c4c9d6;
}

.cache-1w4yazo {
  box-sizing: content-box;
  outline: none;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 24px;
  position: relative;
  -webkit-transition: all 300ms;
  transition: all 300ms;
  background-color: #b2b6c3;
  width: 48px;
  height: 24px;
}

.cache-1w4yazo:after {
  content: '';
  position: absolute;
  top: 4px;
  left: 5px;
  width: 16px;
  height: 16px;
  border-radius: 100px;
  background-color: #ffffff;
  -webkit-transition: all 300ms;
  transition: all 300ms;
}

.cache-1w4yazo:focus-within,
.cache-1w4yazo:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
}

.cache-1w4yazo[data-disabled='false']:active:after {
  width: 22.04px;
}

.cache-1w4yazo[data-checked='true'] {
  color: #151a2d;
  background-color: #4f0599;
}

.cache-1w4yazo[data-checked='true']:after {
  left: calc(100% - 5px);
  -webkit-transform: translateX(-100%);
  -moz-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
}

.cache-1w4yazo[data-checked='true']:focus-within,
.cache-1w4yazo[data-checked='true']:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
}

.cache-1w4yazo[data-disabled='true'] {
  background: #d4dae7;
}

.cache-1w4yazo[data-disabled='true'][data-checked='true'] {
  background: #bda5df;
}

.cache-p9zju0 {
  position: absolute;
  opacity: 0;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  cursor: pointer;
}

.cache-p9zju0[disabled] {
  cursor: not-allowed;
}

<form
    novalidate=""
  >
    <label
      aria-disabled="false"
      class="cache-ou5gie e6hn8de1"
    >
      <div
        class="cache-1w4yazo e6hn8de3"
        data-checked="true"
        data-disabled="false"
      >
        <input
          aria-checked="true"
          aria-label="test"
          checked=""
          class="cache-p9zju0 e6hn8de2"
          id="test"
          name="test"
          type="checkbox"
        />
      </div>
    </label>
  </form>
</DocumentFragment>
`;

exports[`ToggleField should render correctly disabled 1`] = `
<DocumentFragment>
  .cache-ou5gie {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  cursor: pointer;
}

.cache-ou5gie:active .e6hn8de3[data-disabled='false']:after {
  width: 22.04px;
}

.cache-ou5gie[aria-disabled='true'] {
  cursor: not-allowed;
  color: #c4c9d6;
}

.cache-1w4yazo {
  box-sizing: content-box;
  outline: none;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 24px;
  position: relative;
  -webkit-transition: all 300ms;
  transition: all 300ms;
  background-color: #b2b6c3;
  width: 48px;
  height: 24px;
}

.cache-1w4yazo:after {
  content: '';
  position: absolute;
  top: 4px;
  left: 5px;
  width: 16px;
  height: 16px;
  border-radius: 100px;
  background-color: #ffffff;
  -webkit-transition: all 300ms;
  transition: all 300ms;
}

.cache-1w4yazo:focus-within,
.cache-1w4yazo:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
}

.cache-1w4yazo[data-disabled='false']:active:after {
  width: 22.04px;
}

.cache-1w4yazo[data-checked='true'] {
  color: #151a2d;
  background-color: #4f0599;
}

.cache-1w4yazo[data-checked='true']:after {
  left: calc(100% - 5px);
  -webkit-transform: translateX(-100%);
  -moz-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
}

.cache-1w4yazo[data-checked='true']:focus-within,
.cache-1w4yazo[data-checked='true']:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
}

.cache-1w4yazo[data-disabled='true'] {
  background: #d4dae7;
}

.cache-1w4yazo[data-disabled='true'][data-checked='true'] {
  background: #bda5df;
}

.cache-p9zju0 {
  position: absolute;
  opacity: 0;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  cursor: pointer;
}

.cache-p9zju0[disabled] {
  cursor: not-allowed;
}

<form
    novalidate=""
  >
    <label
      aria-disabled="true"
      class="cache-ou5gie e6hn8de1"
    >
      <div
        class="cache-1w4yazo e6hn8de3"
        data-checked="false"
        data-disabled="true"
      >
        <input
          aria-checked="false"
          aria-label="test"
          class="cache-p9zju0 e6hn8de2"
          disabled=""
          id="test"
          name="test"
          type="checkbox"
        />
      </div>
    </label>
  </form>
</DocumentFragment>
`;

exports[`ToggleField should render correctly with label and checked 1`] = `
<DocumentFragment>
  .cache-ou5gie {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  cursor: pointer;
}

.cache-ou5gie:active .e6hn8de3[data-disabled='false']:after {
  width: 22.04px;
}

.cache-ou5gie[aria-disabled='true'] {
  cursor: not-allowed;
  color: #c4c9d6;
}

.cache-1w4yazo {
  box-sizing: content-box;
  outline: none;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 24px;
  position: relative;
  -webkit-transition: all 300ms;
  transition: all 300ms;
  background-color: #b2b6c3;
  width: 48px;
  height: 24px;
}

.cache-1w4yazo:after {
  content: '';
  position: absolute;
  top: 4px;
  left: 5px;
  width: 16px;
  height: 16px;
  border-radius: 100px;
  background-color: #ffffff;
  -webkit-transition: all 300ms;
  transition: all 300ms;
}

.cache-1w4yazo:focus-within,
.cache-1w4yazo:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
}

.cache-1w4yazo[data-disabled='false']:active:after {
  width: 22.04px;
}

.cache-1w4yazo[data-checked='true'] {
  color: #151a2d;
  background-color: #4f0599;
}

.cache-1w4yazo[data-checked='true']:after {
  left: calc(100% - 5px);
  -webkit-transform: translateX(-100%);
  -moz-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
}

.cache-1w4yazo[data-checked='true']:focus-within,
.cache-1w4yazo[data-checked='true']:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
}

.cache-1w4yazo[data-disabled='true'] {
  background: #d4dae7;
}

.cache-1w4yazo[data-disabled='true'][data-checked='true'] {
  background: #bda5df;
}

.cache-p9zju0 {
  position: absolute;
  opacity: 0;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  cursor: pointer;
}

.cache-p9zju0[disabled] {
  cursor: not-allowed;
}

.cache-sg5atl {
  margin-left: 8px;
}

<form
    novalidate=""
  >
    <label
      aria-disabled="false"
      class="cache-ou5gie e6hn8de1"
    >
      <div
        class="cache-1w4yazo e6hn8de3"
        data-checked="true"
        data-disabled="false"
      >
        <input
          aria-checked="true"
          aria-label="test"
          checked=""
          class="cache-p9zju0 e6hn8de2"
          id="test"
          name="test"
          type="checkbox"
        />
      </div>
      <span
        class="cache-sg5atl e6hn8de0"
      >
        test
      </span>
    </label>
  </form>
</DocumentFragment>
`;
